Userland/Applications/Dictionary/Dictionary.cpp

```diff
@@ -0,0 +1,198 @@
+    m_dictionary_bytes = (uint8_t*)m_dictionary_file->data();
+
+    Header const* header = (Header const*)m_dictionary_bytes;
+
+    m_num_words = header->num_words;
+    m_max_word_length = header->max_word_length;
+
+    m_index = offset_to_pointer(header->index_offset);
+    m_word_data_index = offset_to_pointer(header->word_data_index_offset);
+
+    outln("Number of words: {}", m_num_words);
+    outln("Max word length: {}", m_max_word_length);
+
+    VERIFY(m_num_words < 1000 * 1000);
+    VERIFY(m_max_word_length == 16);
+}
+
+StringView Dictionary::word_at_index(int word_index) const
+{
+    if (0 <= word_index && word_index < m_num_words) {
+        char const* c = (char const*)(m_index[word_index]);
+
+        int n = size_of_index_entry(c);
+
+        return StringView(c, n);
+    }
+    return "";
+}
+
+String Dictionary::definition_of(int word_index) const
+{
+    if (!(0 <= word_index && word_index < m_num_words)) {
+        return "";
+    }
+
+    Offset<WordData> word_data_offset = m_word_data_index[word_index];
+
+    WordData const& word_data = *offset_to_pointer(word_data_offset);
+
+    StringBuilder string_builder;
+
+    string_builder.append(word_at_index(word_index));
+    string_builder.append("\n\n");
+
+    char last_part_of_speech = 0;
+    int sense_number = 0;
+
+    int n = word_data.num_senses;
+    for (int i = 0; i < n; i++) {
+        Sense const& sense = word_data.senses[i];
+
+        Definition const& definition = *offset_to_pointer(sense.definition_offset);
+
+        if (definition.part_of_speech != last_part_of_speech) {
+            string_builder.append(part_of_speech_name(definition.part_of_speech));
+            string_builder.append("\n\n");
+
+            last_part_of_speech = definition.part_of_speech;
+            sense_number = 0;
+        }
+
+        if (n > 1)
+            string_builder.append(String::formatted("  {}. ", sense_number + 1));
+
+        string_builder.append(definition.text);
+        string_builder.append("\n\n");
+
+        sense_number += 1;
+    }
+
+    return string_builder.string_view();
+}
+
+void Dictionary::prefix_query(
+    String query,
+    int start_index,
+    int& first_index,
+    int& num_indices) const
+{
+    if (m_num_words == 0) {
+        return;
+    }
+
+    first_index = start_index;
+
+    if (query == "") {
+        num_indices = m_num_words - start_index;
+        return;
+    }
+
+    /* First, find the new first_index by padding the query with zeroes
+           and finding the first entry that's >= it. */
+    char query_chars[16] = {};
+    memcpy(query_chars, query.to_lowercase().characters(), query.length());
+
+    for (int letter_index = 0; letter_index < 16; letter_index++) {
+
+        char query_char = query_chars[letter_index];
+
+        char const* p_item_char = &m_index[first_index][letter_index];
+
+        /* OR with 32 to get the lowercase char */
+        while ((*p_item_char | 32) < query_char) {
```

Comment:
I'd recommend using AK/CharacterTypes.h for this instead of using bitwise math here for clarity.

For example, if you iterate from 0 to min(query.length(), 16) as suggested above, you could do something like
```c++
  auto query_lowercase = query.to_lowercase(); // creates a new String on the heap, fwiw
  const size_t max_letters = min(query_lowercase.length(), 16);
  for (size_t letter_index = 0; letter_index < max_letters; ++letter_index) {
     char query_char = query_chars[letter_index];
     char const* p_item_char = &m_index[first_index][letter_index];
     while (to_ascii_lowercase(*p_item_char) < query_char) {
          ...
```

Reply:
The index only contains ASCII A-Z and a-z, so I don't really love using `to_ascii_lowercase` here, which will do a check that each char in the index is between `'A'` and `'Z'` every time.

I knew the unadorned `| 32` could be unclear, but I think the comment solves that. I suppose I could make it a function instead, but it didn't seem worth it?